File: backend/app/api/v1/hr/attendance.py
from __future__ import annotations

import datetime as dt

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.auth import require_permissions
from app.core.database import get_db
from app.modules.hr_attendance.schemas import (
    AttendanceEntryCreate,
    AttendanceEntryListQuery,
    AttendanceEntryPublic,
)
from app.modules.hr_attendance.service import (
    create_attendance_entry,
    list_attendance_entries,
)

router = APIRouter(prefix="/attendance")


@router.post(
    "",
    response_model=AttendanceEntryPublic,
    dependencies=[Depends(require_permissions({"hr.attendance.write"}))],
)
def hr_create_attendance_entry(
    payload: AttendanceEntryCreate,
    db: Session = Depends(get_db),
) -> AttendanceEntryPublic:
    row = create_attendance_entry(
        db,
        employee_id=payload.employee_id,
        project_id=payload.project_id,
        work_date=payload.work_date,
        hours=payload.hours,
        hourly_rate=payload.hourly_rate,
        notes=payload.notes,
    )
    return AttendanceEntryPublic.model_validate(row)


@router.get(
    "",
    response_model=list[AttendanceEntryPublic],
    dependencies=[Depends(require_permissions({"hr.attendance.read"}))],
)
def hr_list_attendance_entries(
    project_id: int | None = None,
    employee_id: int | None = None,
    date_from: dt.date | None = None,
    date_to: dt.date | None = None,
    limit: int = 50,
    offset: int = 0,
    db: Session = Depends(get_db),
) -> list[AttendanceEntryPublic]:
    q = AttendanceEntryListQuery(
        project_id=project_id,
        employee_id=employee_id,
        date_from=date_from,
        date_to=date_to,
        limit=limit,
        offset=offset,
    )
    rows = list_attendance_entries(
        db,
        project_id=q.project_id,
        employee_id=q.employee_id,
        date_from=q.date_from,
        date_to=q.date_to,
        limit=q.limit,
        offset=q.offset,
    )
    return [AttendanceEntryPublic.model_validate(r) for r in rows]

File: frontend/src/api/hr/attendance.ts
import { apiFetch } from '../client';

export type AttendanceEntry = {
  id: number;
  employee_id: number;
  project_id: number;
  work_date: string;
  hours: number;
  hourly_rate: number;
  notes: string | null;
  created_at: string;
  updated_at: string;
};

export type AttendanceEntryCreate = {
  employee_id: number;
  project_id: number;
  work_date: string;
  hours: number;
  hourly_rate: number;
  notes?: string | null;
};

export type AttendanceEntryListQuery = {
  project_id?: number;
  employee_id?: number;
  date_from?: string;
  date_to?: string;
  limit?: number;
  offset?: number;
};

export function createAttendanceEntry(payload: AttendanceEntryCreate): Promise<AttendanceEntry> {
  return apiFetch<AttendanceEntry>('/api/v1/hr/attendance', {
    method: 'POST',
    body: JSON.stringify({
      ...payload,
      notes: payload.notes ?? null,
    }),
  });
}

export function listAttendanceEntries(q: AttendanceEntryListQuery): Promise<AttendanceEntry[]> {
  const params = new URLSearchParams();
  if (q.project_id != null) params.set('project_id', String(q.project_id));
  if (q.employee_id != null) params.set('employee_id', String(q.employee_id));
  if (q.date_from) params.set('date_from', q.date_from);
  if (q.date_to) params.set('date_to', q.date_to);
  if (q.limit != null) params.set('limit', String(q.limit));
  if (q.offset != null) params.set('offset', String(q.offset));
  const qs = params.toString();
  return apiFetch<AttendanceEntry[]>(`/api/v1/hr/attendance${qs ? `?${qs}` : ''}`);
}
